### PR TITLE
feat: watermark for sources

### LINKED ISSUE #88: NoOp watermark support
use NoOp watermark progressor and avoid nil checks in forwarder.

### PR DESCRIPTION
**Pull request by whynowy**
Signed-off-by: Derek Wang <whynowy@gmail.com>

1. Add wartermark for Kafka and HTTP sources;
2. Add `maxDelay` to watermark config;
3. Enable wartermark by default.

Fixes #88

### CODE REVIEW

File: pkg/watermark/fetch/processor_manager.go [hunk 1]
```diff
@@ -15,21 +15,10 @@ import (
...
-	// DeleteProcessor deletes a processor.
-	DeleteProcessor(processor string)
-}
-
-// fromVertex is the point of view of Vn-1 from Vn vertex. The code is running on Vn vertex.
+// ProcessorManager manages the point of view of Vn-1 from Vn vertex processors (or source processor). The code is running on Vn vertex.
 // It has the mapping of all the processors which in turn has all the information about each processor
 // timelines.
-type fromVertex struct {
+type ProcessorManager struct {
```
Comment (vigith) [inline pkg/watermark/fetch/processor_manager.go:21]: do we need this to be exported?

**Review by vigith — APPROVED** [review]

**Comment by vigith** [comment]
We should remove the `nil` check in the forwarder. That way we can close #88

File: pkg/watermark/fetch/source_fetcher.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			continue
+		}
+		if p.offsetTimeline.GetHeadWatermark() > epoch {
+			epoch = p.offsetTimeline.GetHeadWatermark()
+		}
+	}
+	if epoch == math.MinInt64 {
+		return processor.Watermark(time.Time{})
+	}
+	return processor.Watermark(time.Unix(epoch, 0))
```
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/watermark/fetch/source_fetcher.go:47]: This should be correct.

File: pkg/watermark/fetch/source_fetcher.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+			continue
+		}
+		if p.offsetTimeline.GetHeadWatermark() < epoch {
+			epoch = p.offsetTimeline.GetHeadWatermark()
+		}
+	}
+	if epoch == math.MaxInt64 {
+		return processor.Watermark(time.Time{})
+	}
+	return processor.Watermark(time.Unix(epoch, 0))
```
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/watermark/fetch/source_fetcher.go:65]: This should be also correct.

[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/watermark/fetch/source_fetcher.go:65]: This should be also correct.

File: pkg/watermark/fetch/processor_manager.go (hunk 1 above)
Comment (whynowy) [reply pkg/watermark/fetch/processor_manager.go:21]: yes.

File: pkg/isb/stores/jetstream/writer.go [hunk 4]
```diff
@@ -261,7 +261,7 @@ func (w *writeOffset) AckIt() error {
 const (
 	_key       = "k"
 	_id        = "i"
-	_window    = "w"
+	_late      = "w"
```
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/isb/stores/jetstream/writer.go:264]: done

File: pkg/watermark/generic/publisher.go [hunk 5]
```diff
@@ -23,36 +22,10 @@ func BuildPublishWMStores(hbStore store.WatermarkKVStorer, otStore store.Waterma
...
 
-// genericPublish is a generic publisher which will work for most cases.
-type genericPublish struct {
-	toEdge publish.Publisher
-}
-
-var _ publish.Publisher = (*genericPublish)(nil)
-
 // NewGenericPublish returns GenericPublish. processorName is the unique processor (pod) that is running on this vertex.
 // publishKeyspace is obsolete, and will be removed in subsequent iterations. publishWM is a struct for storing both the heartbeat
```
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/watermark/generic/publisher.go]: done

**Pull request merged**